src/LanguageServer/Microsoft.CommonLanguageServerProtocol.Framework/QueueItem.cs

```diff
@@ -56,59 +52,85 @@ internal class QueueItem<TRequest, TResponse, TRequestContext> : IQueueItem<TReq
 
         return (queueItem, queueItem._completionSource.Task);
     }
 
-    public async Task<TRequestContext> CreateRequestContextAsync(IMethodHandler handler, CancellationToken cancellationToken)
+    public async Task<TRequestContext> CreateRequestContextAsync<TRequest>(TRequest request, IMethodHandler handler, CancellationToken cancellationToken)
     {
         cancellationToken.ThrowIfCancellationRequested();
 
         _requestTelemetryScope?.RecordExecutionStart();
 
         var requestContextFactory = LspServices.GetRequiredService<AbstractRequestContextFactory<TRequestContext>>();
-        var context = await requestContextFactory.CreateRequestContextAsync(this, handler, _request, cancellationToken).ConfigureAwait(false);
+        var context = await requestContextFactory.CreateRequestContextAsync(this, handler, request, cancellationToken).ConfigureAwait(false);
         return context;
     }
 
+    public TRequest? TryDeserializeRequest<TRequest>(AbstractLanguageServer<TRequestContext> languageServer, RequestHandlerMetadata requestHandlerMetadata, bool isMutating)
+    {
+        try
+        {
+            return languageServer.DeserializeRequest<TRequest>(SerializedRequest, requestHandlerMetadata);
+        }
+        catch (Exception ex)
+        {
+            // Report the exception to logs / telemetry
+            _requestTelemetryScope?.RecordException(ex);
+            _logger.LogException(ex);
+
+            // Set the task result to the exception
+            _completionSource.TrySetException(ex);
+
+            // End the request - the caller will return immediately if it cannot deserialize.
+            _requestTelemetryScope?.Dispose();
+            _logger.LogEndContext($"{MethodName}");
+
+            // If the request is mutating, bubble the exception out so the queue shutsdown.
+            if (isMutating)
+            {
+                throw;
+            }
+            else
+            {
+                return default;
+            }
+        }
+    }
+
     /// <summary>
     /// Processes the queued request. Exceptions will be sent to the task completion source
     /// representing the task that the client is waiting for, then re-thrown so that
     /// the queue can correctly handle them depending on the type of request.
     /// </summary>
-    /// <param name="context">Context used for the request.</param>
-    /// <param name="handler">The handler to execute.</param>
-    /// <param name="cancellationToken"></param>
-    /// <returns>The result of the request.</returns>
-    public async Task StartRequestAsync(TRequestContext? context, IMethodHandler handler, CancellationToken cancellationToken)
+    public async Task StartRequestAsync<TRequest, TResponse>(TRequest request, TRequestContext? context, IMethodHandler handler, string language, CancellationToken cancellationToken)
     {
         _logger.LogStartContext($"{MethodName}");
+        _requestTelemetryScope?.UpdateLanguage(language);
```

Comment:
1. doc.  without the convo we just had, this would have looked super wonky to me.
2. i don't *love* that this is done unilaterally for all requests.   i would almost prefer we at least have some asserts that this MUST be a DidOpen request that can change the lang.

Reply:
Will update.  This is not changing the language - this is reporting the handler language to telemetry that we used to handle the request (based on the langauge set by the didOpen).  Will make it clearer